File: Problem/BOJ_3300_무어 기계/main.py
from sys import stdin
import re


def main():
    stdin = open('./input.txt', 'r')
    test_case = int(stdin.readline())

    for _ in range(test_case):
        machine = stdin.readline().strip()
        output = stdin.readline().strip()

        match_count = 0
        match_char = ''
        for code in range(ord('A'), ord('Z') + 1):
            new_machine = machine.replace("_", chr(code))
            regex = re.compile("^" + new_machine + "$")

            if regex.match(output) is not None:
                match_count += 1
                match_char = chr(code)

        if match_count == 0:
            print("!")
        elif match_count == 1:
            print(match_char)
        else:
            print("_")


if __name__ == '__main__':
    main()

File: Problem/BOJ_3300_무어 기계/main.js
const fs = require('fs')
const input = fs.readFileSync("./input.txt").toString().trim().split("\n").map(elem => elem.trim())

const numOfMachines = parseInt(input[0])
const machines = input.slice(1)


for (let i = 0; i < numOfMachines * 2; i += 2) {
    const machine = machines[i]
    const output = machines[i + 1]

    let matchCount = 0
    let matchChar = ''
    for (let i = "A".charCodeAt(0); i <= "Z".charCodeAt(0); i++) {
        const newMachine = machine.replace("_", String.fromCharCode(i))
        const regex = new RegExp(`^${newMachine}$`)

        if (!!output.match(regex)) {
            matchCount += 1
            matchChar = String.fromCharCode(i)
        }
    }

    if (matchCount === 0) {
        console.log("!")
    } else if (matchCount > 1) {
        console.log("_")
    } else {
        console.log(matchChar)
    }
}